onnxruntime/core/optimizer/utils.cc

```diff
@@ -276,19 +276,22 @@ constexpr std::array kOnnxDomainNonDeterministicOps{"RandomUniform", "RandomNorm
 
 #ifdef ENABLE_TRAINING_OPS
 constexpr std::array kMSDomainDeterministicOps{"ShrunkenGather"};
+#else
+// Operator types in the com.microsoft domain that are known to be deterministic.
+constexpr std::array kMSDomainDeterministicOps{"QuantizeLinear", "DequantizeLinear"};
```

Comment:
should these also be in the list if `ENABLE_TRAINING_OPS` is defined?

Reply:
Good question, I'm not entirely sure :). It would be a good idea to get additional reviews from folks that work on training.

Reply:
@askhade Do you know if adding com.microsoft.Quantize/Dequantize ops to the list of deterministic training ops would be fine? Alternatively, should we make this a list of _non-deterministic_ contrib ops to be more consistent with the list of non-deterministic ONNX ops? Thanks for your help!

Reply:
@pengwa looks like this list was added in #15461. would you mind sharing the rationale for it and whether it should be updated here?

Reply:
Adding these ops to the training-specific list is needed to ensure that the constant folding and cse tests added to this PR pass in builds with `ENABLE_TRAINING_OPS` enabled. If this causes issues, I could disable these units tests for training builds, but I suspect these changes should be fine.

Reply:
DeQuantizeLinear should not be added to deterministic ops list. This list is queried by 2 optimizers constant folding and common subexpression elimination. If DQ is added to this list, common sub expression elimination optimizer will coalesce multiple DQ nodes into 1. QDQ handling does not allow DQ node to be in multiple node groups.

Reply:
@askhade and I got on a call and verified that there is logic in place in both CSE and CF to prevent QDQ node groups from being unnecessarily affected. We agreed to add more detailed comments that explain the difference between the onnx list and the ms_domain list (with the goal of making both lists consistently list non-deterministic ops).

Reply:
> @pengwa looks like this list was added in #15461. would you mind sharing the rationale for it and whether it should be updated here?

I think as long as inference has the need for defining ops in the list, we could make it a common list. The reason I add training macro is to I only see training has such requirements, and I don't want to introduce burden for inference builds. So that means I think one list should be fine. Hope it makes sense.

Reply:
Thanks @pengwa, that makes sense. The PR adds a TODO to go back and make the onnx domain and ms domain lists consistent. As you say, perhaps they can be combined.